Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateIOS/MSFT_IntuneTrustedRootCertificateIOS.psm1

```diff
@@ -0,0 +1,726 @@
+        $Assignments,
+        #endregion
+
+        [Parameter()]
+        [System.String]
+        [ValidateSet('Absent', 'Present')]
+        $Ensure = 'Present',
+
+        [Parameter()]
+        [System.Management.Automation.PSCredential]
+        $Credential,
+
+        [Parameter()]
+        [System.String]
+        $ApplicationId,
+
+        [Parameter()]
+        [System.String]
+        $TenantId,
+
+        [Parameter()]
+        [System.Management.Automation.PSCredential]
+        $ApplicationSecret,
+
+        [Parameter()]
+        [System.String]
+        $CertificateThumbprint,
+
+        [Parameter()]
+        [Switch]
+        $ManagedIdentity,
+
+        [Parameter()]
+        [System.String[]]
+        $AccessTokens
+
+    )
+
+    #Ensure the proper dependencies are installed in the current environment.
+    Confirm-M365DSCDependencies
+
+    #region Telemetry
+    $ResourceName = $MyInvocation.MyCommand.ModuleName.Replace('MSFT_', '')
+    $CommandName = $MyInvocation.MyCommand
+    $data = Format-M365DSCTelemetryParameters -ResourceName $ResourceName `
+        -CommandName $CommandName `
+        -Parameters $PSBoundParameters
+    Add-M365DSCTelemetryEvent -Data $data
+    #endregion
+
+    Write-Verbose -Message "Testing configuration of {$id}"
+
+    $CurrentValues = Get-TargetResource @PSBoundParameters
+    $ValuesToCheck = ([Hashtable]$PSBoundParameters).clone()
+
+    if ($CurrentValues.Ensure -ne $Ensure)
+    {
+        Write-Verbose -Message "Test-TargetResource returned $false"
+        return $false
+    }
+    $testResult = $true
+
+    foreach ($key in $PSBoundParameters.Keys)
+    {
+        if ($PSBoundParameters[$key].getType().Name -like '*CimInstance*')
+        {
+            $CIMArraySource = @()
+            $CIMArrayTarget = @()
+            $CIMArraySource += $PSBoundParameters[$key]
+            $CIMArrayTarget += $CurrentValues.$key
+            if ($CIMArraySource.count -ne $CIMArrayTarget.count)
+            {
+            Write-Verbose -Message "$PSBoundParameters[$key]"
+                Write-Verbose -Message "Configuration drift:Number of items does not match: Source=$($CIMArraySource.count) Target=$($CIMArrayTarget.count)"
+                $testResult = $false
+                break
+            }
+            $i = 0
+            foreach ($item in $CIMArraySource )
+            {
+                $testResult = Compare-M365DSCComplexObject `
+                    -Source (Get-M365DSCDRGComplexTypeToHashtable -ComplexObject $CIMArraySource[$i]) `
+                    -Target ($CIMArrayTarget[$i])
+
+                $i++
+                if (-Not $testResult)
+                {
+                    $testResult = $false
+                    break
+                }
+            }
+            if (-Not $testResult)
+            {
+                $testResult = $false
+                break
+            }
+
+            $ValuesToCheck.Remove($key) | Out-Null
+        }
+    }
```

Comment:
Why complicate things here? Compare-M365DSCComplexObject can compare an whole object, even if it's an array, you don't need to iterate it. It also can compare the number of objects of the source and target, you don't need to do it manually.

Again see https://github.com/microsoft/Microsoft365DSC/blob/5ec749efe0485874e832b85b5d9a77fb37a07e3a/Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationDevicePolicy/MSFT_IntuneAppConfigurationDevicePolicy.psm1#L600 for reference, this chunk is used quite literally in all Intune resources in *Test-TargetResource*.

Reply:
OK, I'll make the edit. It isn't that way in MSFT_IntuneWifiConfigurationPolicyWindows10 either (which is the one I randomly chose to base mine on!)

Reply:
Not your fault but the Wifi resources are not a good example since they don't work correctly and have not been maintained for a long time.

Reply:
Haha, trust me to pick that then.

Reply:
done